// @rviscomi/capo.js/dist/adapters/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/adapters/index.js
var adapters_exports = {};
__export(adapters_exports, {
  AdapterInterface: () => AdapterInterface,
  BrowserAdapter: () => BrowserAdapter,
  validateAdapter: () => validateAdapter
});
module.exports = __toCommonJS(adapters_exports);

// src/adapters/adapter.js
var REQUIRED_METHODS = [
  "isElement",
  "getTagName",
  "getAttribute",
  "hasAttribute",
  "getAttributeNames",
  "getTextContent",
  "getChildren",
  "getParent",
  "getSiblings",
  "stringify"
];
var AdapterInterface = class {
  /**
   * Check if node is an Element (not text, comment, etc.)
   * @param {any} node - The node to check
   * @returns {boolean}
   */
  isElement(node) {
    throw new Error("isElement() not implemented");
  }
  /**
   * Get the tag name of an element (lowercase)
   * @param {any} node - Element node
   * @returns {string} - Tag name like 'meta', 'link', 'script'
   */
  getTagName(node) {
    throw new Error("getTagName() not implemented");
  }
  /**
   * Get attribute value from element
   * @param {any} node - Element node
   * @param {string} attrName - Attribute name (case-insensitive)
   * @returns {string | null} - Attribute value or null if not found
   */
  getAttribute(node, attrName) {
    throw new Error("getAttribute() not implemented");
  }
  /**
   * Check if element has a specific attribute
   * @param {any} node - Element node
   * @param {string} attrName - Attribute name (case-insensitive)
   * @returns {boolean} - True if attribute exists
   */
  hasAttribute(node, attrName) {
    throw new Error("hasAttribute() not implemented");
  }
  /**
   * Get all attribute names for an element
   * @param {any} node - Element node
   * @returns {string[]} - Array of attribute names
   */
  getAttributeNames(node) {
    throw new Error("getAttributeNames() not implemented");
  }
  /**
   * Get text content of a node (for inline scripts/styles)
   * @param {any} node - Element node
   * @returns {string} - Text content
   */
  getTextContent(node) {
    throw new Error("getTextContent() not implemented");
  }
  /**
   * Get child elements of a node
   * @param {any} node - Parent node
   * @returns {any[]} - Array of child element nodes (excluding text/comment nodes)
   */
  getChildren(node) {
    throw new Error("getChildren() not implemented");
  }
  /**
   * Get parent element of a node
   * @param {any} node - Child node
   * @returns {any | null} - Parent element node, or null if no parent
   */
  getParent(node) {
    throw new Error("getParent() not implemented");
  }
  /**
   * Get sibling elements of a node
   * @param {any} node - Element node
   * @returns {any[]} - Array of sibling element nodes (excluding the node itself)
   */
  getSiblings(node) {
    throw new Error("getSiblings() not implemented");
  }
  /**
   * Get source location for a node (optional, for linting)
   * @param {any} node - Element node
   * @returns {{ line: number, column: number, endLine?: number, endColumn?: number } | null}
   */
  getLocation(node) {
    return null;
  }
  /**
   * Stringify element for logging/debugging
   * @param {any} node - Element node
   * @returns {string} - String representation like "<meta charset='utf-8'>"
   */
  stringify(node) {
    throw new Error("stringify() not implemented");
  }
};
function validateAdapter(adapter) {
  for (const method of REQUIRED_METHODS) {
    if (typeof adapter[method] !== "function") {
      throw new Error(`Adapter missing required method: ${method}()`);
    }
  }
}

// src/adapters/browser.js
var BrowserAdapter = class extends AdapterInterface {
  /**
   * Check if node is an Element (not text, comment, etc.)
   * @param {any} node - The node to check
   * @returns {boolean}
   */
  isElement(node) {
    if (!node) return false;
    return node.nodeType === 1;
  }
  /**
   * Get the tag name of an element (lowercase)
   * @param {any} node - Element node
   * @returns {string} - Tag name like 'meta', 'link', 'script'
   */
  getTagName(node) {
    if (!node || !node.tagName) {
      return "";
    }
    return node.tagName.toLowerCase();
  }
  /**
   * Get attribute value from element
   * @param {any} node - Element node
   * @param {string} attrName - Attribute name (case-insensitive)
   * @returns {string | null} - Attribute value or null if not found
   */
  getAttribute(node, attrName) {
    if (!node || typeof node.getAttribute !== "function") {
      return null;
    }
    return node.getAttribute(attrName);
  }
  /**
   * Check if element has a specific attribute
   * @param {any} node - Element node
   * @param {string} attrName - Attribute name (case-insensitive)
   * @returns {boolean} - True if attribute exists
   */
  hasAttribute(node, attrName) {
    if (!node || typeof node.hasAttribute !== "function") {
      return false;
    }
    return node.hasAttribute(attrName);
  }
  /**
   * Get all attribute names for an element
   * @param {any} node - Element node
   * @returns {string[]} - Array of attribute names
   */
  getAttributeNames(node) {
    if (!node || typeof node.getAttributeNames !== "function") {
      return [];
    }
    return node.getAttributeNames();
  }
  /**
   * Get text content of a node (for inline scripts/styles)
   * @param {any} node - Element node
   * @returns {string} - Text content
   */
  getTextContent(node) {
    if (!node) {
      return "";
    }
    return node.textContent || "";
  }
  /**
   * Get child elements of a node
   * @param {any} node - Parent node
   * @returns {any[]} - Array of child element nodes (excluding text/comment nodes)
   */
  getChildren(node) {
    if (!node || !node.children) {
      return [];
    }
    return Array.from(node.children);
  }
  /**
   * Get parent element of a node
   * @param {any} node - Child node
   * @returns {any | null} - Parent element node, or null if no parent
   */
  getParent(node) {
    if (!node) {
      return null;
    }
    return node.parentElement || null;
  }
  /**
   * Get sibling elements of a node
   * @param {any} node - Element node
   * @returns {any[]} - Array of sibling element nodes (excluding the node itself)
   */
  getSiblings(node) {
    if (!node) {
      return [];
    }
    const parent = node.parentElement;
    if (!parent) {
      return [];
    }
    return Array.from(parent.children).filter((child) => child !== node);
  }
  /**
   * Get source location for a node (optional, for linting)
   * 
   * Browser DOM elements don't have source location information,
   * so this always returns null.
   * 
   * @param {any} node - Element node
   * @returns {null}
   */
  getLocation(node) {
    return null;
  }
  /**
   * Stringify element for logging/debugging
   * @param {any} node - Element node
   * @returns {string} - String representation like "<meta charset='utf-8'>"
   */
  stringify(node) {
    if (!node || !node.nodeName) {
      return "[invalid node]";
    }
    const tagName = this.getTagName(node);
    const attrNames = this.getAttributeNames(node);
    if (attrNames.length === 0) {
      return `<${tagName}>`;
    }
    const attrs = attrNames.map((attr) => {
      const value = this.getAttribute(node, attr);
      const escapedValue = value ? value.replace(/"/g, "&quot;") : "";
      return `${attr}="${escapedValue}"`;
    }).join(" ");
    return `<${tagName} ${attrs}>`;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AdapterInterface,
  BrowserAdapter,
  validateAdapter
});
