// @rviscomi/capo.js/dist/lib/colors.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/colors.js
var colors_exports = {};
__export(colors_exports, {
  BLUE: () => BLUE,
  DEFAULT: () => DEFAULT,
  PINK: () => PINK,
  Palettes: () => Palettes,
  generateSwatches: () => generateSwatches,
  getInvalidBackgroundColor: () => getInvalidBackgroundColor
});
module.exports = __toCommonJS(colors_exports);
var Hues = {
  PINK: 320,
  BLUE: 200
};
function generateSwatches(hue) {
  return [
    `oklch(5% .1 ${hue})`,
    `oklch(13% .2 ${hue})`,
    `oklch(25% .2 ${hue})`,
    `oklch(35% .25 ${hue})`,
    `oklch(50% .27 ${hue})`,
    `oklch(67% .31 ${hue})`,
    `oklch(72% .25 ${hue})`,
    `oklch(80% .2 ${hue})`,
    `oklch(90% .1 ${hue})`,
    `oklch(99% .05 ${hue})`,
    "#ccc"
  ];
}
var DEFAULT = [
  "#9e0142",
  "#d53e4f",
  "#f46d43",
  "#fdae61",
  "#fee08b",
  "#e6f598",
  "#abdda4",
  "#66c2a5",
  "#3288bd",
  "#5e4fa2",
  "#cccccc"
];
var PINK = generateSwatches(Hues.PINK);
var BLUE = generateSwatches(Hues.BLUE);
var Palettes = {
  DEFAULT,
  PINK,
  BLUE
};
function getInvalidBackgroundColor(elementColor) {
  let invalidColor = "#cccccc";
  if (elementColor == invalidColor) {
    invalidColor = "red";
  }
  return `repeating-linear-gradient(45deg, ${elementColor}, ${elementColor} 3px, ${invalidColor} 3px, ${invalidColor} 6px)`;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BLUE,
  DEFAULT,
  PINK,
  Palettes,
  generateSwatches,
  getInvalidBackgroundColor
});
