// @rviscomi/capo.js/dist/lib/options.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/options.js
var options_exports = {};
__export(options_exports, {
  Options: () => Options
});
module.exports = __toCommonJS(options_exports);

// src/lib/colors.js
var Hues = {
  PINK: 320,
  BLUE: 200
};
function generateSwatches(hue) {
  return [
    `oklch(5% .1 ${hue})`,
    `oklch(13% .2 ${hue})`,
    `oklch(25% .2 ${hue})`,
    `oklch(35% .25 ${hue})`,
    `oklch(50% .27 ${hue})`,
    `oklch(67% .31 ${hue})`,
    `oklch(72% .25 ${hue})`,
    `oklch(80% .2 ${hue})`,
    `oklch(90% .1 ${hue})`,
    `oklch(99% .05 ${hue})`,
    "#ccc"
  ];
}
var DEFAULT = [
  "#9e0142",
  "#d53e4f",
  "#f46d43",
  "#fdae61",
  "#fee08b",
  "#e6f598",
  "#abdda4",
  "#66c2a5",
  "#3288bd",
  "#5e4fa2",
  "#cccccc"
];
var PINK = generateSwatches(Hues.PINK);
var BLUE = generateSwatches(Hues.BLUE);
var Palettes = {
  DEFAULT,
  PINK,
  BLUE
};

// src/lib/options.js
var Options = class _Options {
  constructor({
    preferredAssessmentMode = _Options.AssessmentMode.STATIC,
    validation = true,
    palette = DEFAULT,
    loggingPrefix = "Capo: "
  } = {}) {
    this.setPreferredAssessmentMode(preferredAssessmentMode);
    this.setValidation(validation);
    this.setPalette(palette);
    this.setLoggingPrefix(loggingPrefix);
  }
  static get AssessmentMode() {
    return {
      STATIC: "static",
      DYNAMIC: "dynamic"
    };
  }
  static get Palettes() {
    return Palettes;
  }
  prefersStaticAssessment() {
    return this.preferredAssessmentMode === _Options.AssessmentMode.STATIC;
  }
  prefersDynamicAssessment() {
    return this.preferredAssessmentMode === _Options.AssessmentMode.DYNAMIC;
  }
  isValidationEnabled() {
    return this.validation;
  }
  setPreferredAssessmentMode(preferredAssessmentMode) {
    if (!this.isValidAssessmentMode(preferredAssessmentMode)) {
      throw new Error(`Invalid option: preferred assessment mode, expected AssessmentMode.STATIC or AssessmentMode.DYNAMIC, got "${preferredAssessmentMode}".`);
    }
    this.preferredAssessmentMode = preferredAssessmentMode;
  }
  setPreferredAssessmentModeToStatic(prefersStatic) {
    let mode = _Options.AssessmentMode.STATIC;
    if (!prefersStatic) {
      mode = _Options.AssessmentMode.DYNAMIC;
    }
    this.setPreferredAssessmentMode(mode);
  }
  setValidation(validation) {
    if (!this.isValidValidation(validation)) {
      throw new Error(`Invalid option: validation, expected boolean, got "${validation}".`);
    }
    this.validation = validation;
  }
  setPalette(palette) {
    if (!this.isValidPalette(palette)) {
      throw new Error(`Invalid option: palette, expected [${Object.keys(Palettes).join("|")}] or an array of colors, got "${palette}".`);
    }
    if (typeof palette === "string") {
      this.palette = Palettes[palette];
      return;
    }
    this.palette = palette;
  }
  setLoggingPrefix(loggingPrefix) {
    if (!this.isValidLoggingPrefix(loggingPrefix)) {
      throw new Error(`Invalid option: logging prefix, expected string, got "${loggingPrefix}".`);
    }
    this.loggingPrefix = loggingPrefix;
  }
  isValidAssessmentMode(assessmentMode) {
    return Object.values(_Options.AssessmentMode).includes(assessmentMode);
  }
  isValidValidation(validation) {
    return typeof validation === "boolean";
  }
  isValidPalette(palette) {
    if (typeof palette === "string") {
      return Object.keys(Palettes).includes(palette);
    }
    if (!Array.isArray(palette)) {
      return false;
    }
    return palette.length === 11 && palette.every((color) => typeof color === "string");
  }
  isValidLoggingPrefix(loggingPrefix) {
    return typeof loggingPrefix === "string";
  }
  isPreferredPalette(palette) {
    return JSON.stringify(this.palette) == JSON.stringify(palette);
  }
  valueOf() {
    return {
      preferredAssessmentMode: this.preferredAssessmentMode,
      validation: this.validation,
      palette: this.palette,
      loggingPrefix: this.loggingPrefix
    };
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Options
});
