// @rviscomi/capo.js/dist/lib/rules.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib/rules.js
var rules_exports = {};
__export(rules_exports, {
  ElementDetectors: () => ElementDetectors,
  ElementWeights: () => ElementWeights,
  META_HTTP_EQUIV_KEYWORDS: () => META_HTTP_EQUIV_KEYWORDS,
  getHeadWeights: () => getHeadWeights,
  getWeight: () => getWeight,
  isAsyncScript: () => isAsyncScript,
  isDeferScript: () => isDeferScript,
  isImportStyles: () => isImportStyles,
  isMeta: () => isMeta,
  isMetaCSP: () => isMetaCSP,
  isOriginTrial: () => isOriginTrial,
  isPreconnect: () => isPreconnect,
  isPrefetchPrerender: () => isPrefetchPrerender,
  isPreload: () => isPreload,
  isSyncScript: () => isSyncScript,
  isSyncStyles: () => isSyncStyles,
  isTitle: () => isTitle
});
module.exports = __toCommonJS(rules_exports);
var ElementWeights = {
  META: 10,
  TITLE: 9,
  PRECONNECT: 8,
  ASYNC_SCRIPT: 7,
  IMPORT_STYLES: 6,
  SYNC_SCRIPT: 5,
  SYNC_STYLES: 4,
  PRELOAD: 3,
  DEFER_SCRIPT: 2,
  PREFETCH_PRERENDER: 1,
  OTHER: 0
};
var ElementDetectors = {
  META: isMeta,
  TITLE: isTitle,
  PRECONNECT: isPreconnect,
  ASYNC_SCRIPT: isAsyncScript,
  IMPORT_STYLES: isImportStyles,
  SYNC_SCRIPT: isSyncScript,
  SYNC_STYLES: isSyncStyles,
  PRELOAD: isPreload,
  DEFER_SCRIPT: isDeferScript,
  PREFETCH_PRERENDER: isPrefetchPrerender
};
var META_HTTP_EQUIV_KEYWORDS = [
  "accept-ch",
  "content-security-policy",
  "content-type",
  "default-style",
  "delegate-ch",
  "origin-trial",
  "x-dns-prefetch-control"
];
function isMeta(element, adapter) {
  const tagName = adapter.getTagName(element);
  if (tagName === "base") {
    return true;
  }
  if (tagName !== "meta") {
    return false;
  }
  if (adapter.hasAttribute(element, "charset")) {
    return true;
  }
  const name = adapter.getAttribute(element, "name");
  if (name && name.toLowerCase() === "viewport") {
    return true;
  }
  const httpEquiv = adapter.getAttribute(element, "http-equiv");
  if (httpEquiv) {
    const normalizedValue = httpEquiv.toLowerCase();
    return META_HTTP_EQUIV_KEYWORDS.includes(normalizedValue);
  }
  return false;
}
function isTitle(element, adapter) {
  return adapter.getTagName(element) === "title";
}
function isPreconnect(element, adapter) {
  if (adapter.getTagName(element) !== "link") {
    return false;
  }
  const rel = adapter.getAttribute(element, "rel");
  return rel?.toLowerCase() === "preconnect";
}
function isAsyncScript(element, adapter) {
  return adapter.getTagName(element) === "script" && adapter.hasAttribute(element, "src") && adapter.hasAttribute(element, "async");
}
function isImportStyles(element, adapter) {
  const importRe = /@import/;
  if (adapter.getTagName(element) === "style") {
    return importRe.test(adapter.getTextContent(element));
  }
  return false;
}
function isSyncScript(element, adapter) {
  if (adapter.getTagName(element) !== "script") {
    return false;
  }
  if (adapter.hasAttribute(element, "src") && adapter.hasAttribute(element, "defer")) {
    return false;
  }
  if (adapter.hasAttribute(element, "src")) {
    const type2 = adapter.getAttribute(element, "type");
    if (type2 && type2.toLowerCase() === "module") {
      return false;
    }
  }
  if (adapter.hasAttribute(element, "src") && adapter.hasAttribute(element, "async")) {
    return false;
  }
  const type = adapter.getAttribute(element, "type");
  if (type && type.toLowerCase().includes("json")) {
    return false;
  }
  return true;
}
function isSyncStyles(element, adapter) {
  const tagName = adapter.getTagName(element);
  if (tagName === "style") {
    return true;
  }
  if (tagName === "link") {
    const rel = adapter.getAttribute(element, "rel");
    return rel?.toLowerCase() === "stylesheet";
  }
  return false;
}
function isPreload(element, adapter) {
  if (adapter.getTagName(element) !== "link") {
    return false;
  }
  const rel = adapter.getAttribute(element, "rel");
  if (!rel) {
    return false;
  }
  const relLower = rel.toLowerCase();
  return relLower === "preload" || relLower === "modulepreload";
}
function isDeferScript(element, adapter) {
  if (adapter.getTagName(element) !== "script") {
    return false;
  }
  if (!adapter.hasAttribute(element, "src")) {
    return false;
  }
  if (adapter.hasAttribute(element, "defer")) {
    return true;
  }
  const type = adapter.getAttribute(element, "type");
  if (type && type.toLowerCase() === "module") {
    return !adapter.hasAttribute(element, "async");
  }
  return false;
}
function isPrefetchPrerender(element, adapter) {
  if (adapter.getTagName(element) !== "link") {
    return false;
  }
  const rel = adapter.getAttribute(element, "rel");
  if (!rel) {
    return false;
  }
  const relLower = rel.toLowerCase();
  return relLower === "prefetch" || relLower === "dns-prefetch" || relLower === "prerender";
}
function isOriginTrial(element, adapter) {
  if (adapter.getTagName(element) !== "meta") {
    return false;
  }
  const httpEquiv = adapter.getAttribute(element, "http-equiv");
  return httpEquiv?.toLowerCase() === "origin-trial";
}
function isMetaCSP(element, adapter) {
  if (adapter.getTagName(element) !== "meta") {
    return false;
  }
  const httpEquiv = adapter.getAttribute(element, "http-equiv");
  if (!httpEquiv) {
    return false;
  }
  const httpEquivLower = httpEquiv.toLowerCase();
  return httpEquivLower === "content-security-policy" || httpEquivLower === "content-security-policy-report-only";
}
function getWeight(element, adapter) {
  for (let [id, detector] of Object.entries(ElementDetectors)) {
    if (detector(element, adapter)) {
      return ElementWeights[id];
    }
  }
  return ElementWeights.OTHER;
}
function getHeadWeights(head, adapter) {
  const headChildren = adapter.getChildren(head);
  return headChildren.filter((element) => {
    const tagName = adapter.getTagName(element);
    return tagName && tagName !== "";
  }).map((element) => {
    return {
      element,
      weight: getWeight(element, adapter)
    };
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ElementDetectors,
  ElementWeights,
  META_HTTP_EQUIV_KEYWORDS,
  getHeadWeights,
  getWeight,
  isAsyncScript,
  isDeferScript,
  isImportStyles,
  isMeta,
  isMetaCSP,
  isOriginTrial,
  isPreconnect,
  isPrefetchPrerender,
  isPreload,
  isSyncScript,
  isSyncStyles,
  isTitle
});
